llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp

```diff
@@ -4069,6 +4069,74 @@ SDValue AMDGPUTargetLowering::splitBinaryBitConstantOpImpl(
   return DAG.getNode(ISD::BITCAST, SL, MVT::i64, Vec);
 }
 
```

Comment:
Instead of calling getShiftForReduction in each shift combine function, consider adding a new performExtractVectorElementCombine function that transforms:

`(extract_vector_element (and {y0, y1},  (build_vector 0x1f, 0x1f)), index)`

to:

`(and (extract_vector_element {y0, y1}, index), 0x1f)`

Perhaps this could be generalized to other bitwise operations, other constants, and other vector lengths.